core/partitioning/shape_analysis.cpp

```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
   return ivalue_map;
 }
 
+torch::jit::Node* getUpstreamCastNode(torch::jit::Value* val) {
+  std::queue<torch::jit::Value*> q;
+  q.push(val);
+  std::unordered_set<torch::jit::Node*> visited;
+  while (!q.empty()) {
+    auto cur_val = q.front();
+    q.pop();
+    auto node = cur_val->node();
```

Comment:
For BART (https://huggingface.co/facebook/bart-base), the segfault occurs here, when the `cur_val` pointer is accessed. When `getSegmentsOutputByRunning` calls `createCastNode` on line 221, the index passed in is a `seg_blocks.output()` index, but on line 83, `createCastNode` uses that index for `seg_block.raw_inputs()`. For the block in question, the number of inputs and outputs differs, so this may contribute to the segfault.